common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,243 @@
+     * credits. In this case, the {@link #getUsedCredits()} will exceed the max allowed credits and no other (non-forced)
+     * task will be allowed to execute until {@link #getUsedCredits()} falls below the max allowed.
+     *
+     * A task will allocate the requested credits when it is triggered. If the task fails (synchronously or asynchronously),
+     * then the requested credits are automatically released back into the pool. If the task succeeds, the credits will
+     * remain.
+     *
+     * @param task    A {@link Supplier} that, when invoked, will execute the task.
+     * @param credits The number of credits this task requires.
+     * @param force   If true, the task will be executed synchronously regardless of how many credits are available. The
+     *                task's credits are still recorded in this case.
+     * @param <T>     Return type.
+     * @return A CompletableFuture that, when completed, will contain the result of the executed task. If the task failed
+     * or was rejected (i.e., due to {@link AsyncSemaphore} closing), it will be failed with the appropriate exception.
+     */
+    public <T> CompletableFuture<T> run(@NonNull Supplier<CompletableFuture<T>> task, long credits, boolean force) {
+        Preconditions.checkArgument(credits >= 0 && credits <= this.totalCredits,
+                "credits must be a non-negative number smaller than or equal to %s.", this.totalCredits);
+
+        PendingTask<T> pt;
+        synchronized (this.queue) {
+            Exceptions.checkNotClosed(this.closed, this);
+            if (force || canExecute(credits)) {
+                pt = null;
+                this.usedCredits += credits;
+                log.trace("AsyncSemaphore[{}]: Task run. Credits={}, TotalUsedCredits={}, Forced={}.", this.logId, credits, this.usedCredits, force);
+            } else {
+                // Insufficient credits; need to queue up and execute when more becomes available.
+                pt = new PendingTask<>(credits, task);
+                this.queue.addLast(pt);
+                log.debug("AsyncSemaphore[{}]: Task blocked. Credits={}, TotalUsedCredits={}, QueueSize={}.",
+                        this.logId, credits, this.usedCredits, this.queue.size());
+            }
+        }
+
+        if (pt == null) {
+            // We have more credits than what this task requires. Execute now without queuing.
+            return execute(task, credits);
+        } else {
+            // This wil be completed when its associated task is executed.
+            return pt.result;
+        }
+    }
+
+    /**
+     * Releases a number of credits back into the pool and initiates the execution of any pending tasks that are now
+     * eligible to run.
+     *
+     * @param credits The number of credits to release. This number will be capped at the number of currently used
+     *                credits ({@link #getUsedCredits()}).
+     */
+    public void release(long credits) {
```

Comment:
It would be good if we could automate calling this so that the caller cannot forget. IE by chaining it's invocation on the future returned for example.